File: contents/gaussian_elimination/code/javascript/gaussian_elimination.js
function gaussianElimination(a) {
  const rows = a.length
  const cols = a[0].length
  let row = 0;
  for (let col = 0; col < cols - 1; ++col) {

    let pivot = row;
    for (let i = row + 1; i < rows; ++i) {
      if (Math.abs(a[i][col]) > Math.abs(a[pivot][col])) {
        pivot = i;
      }
    }

    if (a[pivot][col] === 0) {
      console.log("The matrix is singular.");
      continue;
    }

    if (col !== pivot) {
      const t = a[col];
      a[col] = a[pivot];
      a[pivot] = t;
    }

    for (let i = row + 1; i < rows; ++i) {
      const scale = a[i][col] / a[row][col];

      for (let j = col + 1; j < cols; ++j) {
        a[i][j] -= a[row][j] * scale;
      }

      a[i][col] = 0;
    }

    ++row;
  }
  return a;
}

function backSubstitution(a) {
  const rows = a.length;
  const cols = a[0].length;
  const sol = [];

  for (let i = rows - 1; i >= 0; --i) {

    let sum = 0;
    for (let j = cols - 2; j > i; --j) {
      sum += sol[j] * a[i][j];
    }

    sol[i] = (a[i][cols - 1] - sum) / a[i][i];
  }
  return sol;
}

function gaussJordan(a) {
  const cols = a[0].length;
  let row = 0;

  for (let col = 0; col < cols - 1; ++col) {
    if (a[row][col] !== 0) {
      for (let i = cols - 1; i > col - 1; --i) {
        a[row][i] /= a[row][col];
      }

      for (let i = 0; i < row; ++i) {
        for (let j = cols - 1; j > col - 1; --j) {
          a[i][j] -= a[i][col] * a[row][j];
        }
      }

      ++row;
    }
  }
}

function printMatrixRow(row) {
  const text = row
    .map(v => (v < 0 ? " " : "  ") + v.toPrecision(8))
    .join("");

  console.log(text);
}

function printMatrix(a) {
  for (const row of a) {
    printMatrixRow(row);
  }
}

const a = [
  [3,  2, -4,  3],
  [2,  3,  3, 15],
  [5, -3,  1, 14]
];

gaussianElimination(a);
console.log("Gaussian elimination:");
printMatrix(a);

gaussJordan(a);
console.log("\nGauss-Jordan:");
printMatrix(a);

const sol = backSubstitution(a);
console.log("\nSolutions are:");
printMatrixRow(sol);

File: contents/gaussian_elimination/code/go/gaussian_elimination.go
// Package demonstrates Gaussian Elimination
package main

import (
	"fmt"
	"math"
)

func gaussianElimination(a [][]float64) {
	singular := false
	rows := len(a)
	cols := len(a[0])

	for c, r := 0, 0; c < cols && r < rows; c++ {
		// 1. find highest value in column below row to be pivot
		p, highest := r, 0.
		for i, row := range a[r:] {
			if abs := math.Abs(row[c]); abs > highest {
				p = r + i
				highest = abs
			}
		}
		highest = a[p][c] // correct sign

		if highest == 0. {
			if !singular {
				singular = true
				fmt.Println("This matrix is singular.")
			}
			continue
		}

		// 2. swap pivot with current row
		if p != r {
			a[r], a[p] = a[p], a[r]
		}

		for _, row := range a[r+1:] {
			// 3. find fraction from pivot value
			frac := row[c] / highest

			// 4. subtract row to set rest of column to zero
			for j := range row {
				row[j] -= frac * a[r][j]
			}

			// 5. ensure col goes to zero (no float rounding)
			row[c] = 0.
		}

		r++
	}
}

func gaussJordan(a [][]float64) {
	for r := len(a) - 1; r >= 0; r-- {
		// Find pivot col
		p := -1
		for c, cell := range a[r] {
			if cell != 0. {
				p = c
				break
			}
		}
		if p < 0 {
			continue
		}

		// Scale pivot r to 1.
		scale := a[r][p]
		for c := range a[r][p:] {
			a[r][p+c] /= scale
		}
		// Subtract pivot row from each row above
		for _, row := range a[:r] {
			scale = row[p]
			for c, cell := range a[r][p:] {
				row[p+c] -= cell * scale
			}
		}
	}
}

func backSubstitution(a [][]float64) []float64 {
	rows := len(a)
	cols := len(a[0])
	x := make([]float64, rows)
	for r := rows - 1; r >= 0; r-- {
		sum := 0.

		for c := cols - 2; c > r; c-- {
			sum += x[c] * a[r][c]
		}

		x[r] = (a[r][cols-1] - sum) / a[r][r]
	}
	return x
}

func printMatrixRow(row []float64) {
	fmt.Print("[")
	for _, cell := range row {
		fmt.Printf("%9.4f ", cell)
	}
	fmt.Println("]")
}

func printMatrix(a [][]float64) {
	for _, row := range a {
		printMatrixRow(row)
	}
	fmt.Println()
}

func main() {
	a := [][]float64{
		{2, 3, 4, 6},
		{1, 2, 3, 4},
		{3, -4, 0, 10},
	}
	fmt.Println("Original Matrix:")
	printMatrix(a)

	fmt.Println("Gaussian elimination:")
	gaussianElimination(a)
	printMatrix(a)

	gaussJordan(a)
	fmt.Println("Gauss-Jordan:")
	printMatrix(a)

	fmt.Println("Solutions are:")
	x := backSubstitution(a)
	printMatrixRow(x)
}
